### PR TITLE
Update loading_spinner.tsx from class based to functional based

### LINKED ISSUE #24771: Convert ./components/widgets/loading/loading_spinner.tsx from Class Component to Function Component
We want to move most of our codebase from Class Components to Function Components. This decision is based on [React recommendations](https://react.dev/reference/react/Component).

The task is to update the component to be a functional component, without any difference in functionality. Also, all related tests that fail may need refactoring.

Here are a few examples of the required changes:
https://github.com/mattermost/mattermost/pull/24187
https://github.com/mattermost/mattermost/pull/24210
https://github.com/mattermost/mattermost/pull/24183
https://github.com/mattermost/mattermost/pull/24181

If the tests do not work because they rely on the behavior of Class Components they also have to be migrated to functionaly test the same behavior. We prefer the use of `react-testing-library` for new tests, but it is not required.

When submitting the PR, please make sure to identify where and when the component shows, for QA to properly test the changes.

----

If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

JIRA LINK: MM-54808

### PR DESCRIPTION
**Pull request by Krishnakalani111**
#### Summary
Converted ./components/widgets/loading/loading_spinner.tsx from Class Component to Function Component #24771

#### Ticket Link
Fixes #24771 

#### Release Note

```release-note
Converted ./components/widgets/loading/loading_spinner.tsx from Class Component to Function Component #24771
```

### CODE REVIEW

File: webapp/channels/src/components/widgets/loading/loading_spinner.tsx [hunk 1]
```diff
@@ -12,26 +12,26 @@ type Props = {
     style?: React.CSSProperties;
 }
 
-export default class LoadingSpinner extends React.PureComponent<Props> {
-    public static defaultProps: Props = {
-        text: null,
-    };
+const LoadingSpinner: React.FC<Props> = React.memo(({text, style}) => {
```
Comment (jespino) [inline webapp/channels/src/components/widgets/loading/loading_spinner.tsx:15]: nit: We normally put the `React.memo` at the export line, leave it cleaner.

File: webapp/channels/src/components/widgets/loading/loading_spinner.tsx [hunk 2]
```diff
@@ -12,26 +12,26 @@ type Props = {
...
-                    className='fa fa-spinner fa-fw fa-pulse spinner'
-                    component='span'
-                    title={{id: t('generic_icons.loading'), defaultMessage: 'Loading Icon'}}
-                />
-                {this.props.text}
-            </span>
-        );
-    }
-}
+LoadingSpinner.defaultProps = {
```
Comment (jespino) [inline webapp/channels/src/components/widgets/loading/loading_spinner.tsx:33]: The `defaultProps` is not the recomended approach for functional components, is prefered to use the default value for parameters, something like: `const LoadingSpinner: React.FC<Props> = ({text = null, style}) => {`

**Review by jespino — CHANGES_REQUESTED** [review]
A couple of comments, but otherwise looks good to me. Thanks @Krishnakalani111! :woman_dancing:

**Comment by Krishnakalani111** [comment]
Ok I will do the require changes👍

**Comment by Krishnakalani111** [comment]
I have done the required changes please review it.

**Review by jespino — APPROVED** [review]
Nice! Thanks @Krishnakalani111! :tada:

File: webapp/channels/src/components/widgets/loading/loading_spinner.tsx [hunk 3]
```diff
@@ -12,26 +12,22 @@ type Props = {
     style?: React.CSSProperties;
 }
 
-export default class LoadingSpinner extends React.PureComponent<Props> {
-    public static defaultProps: Props = {
-        text: null,
-    };
+const LoadingSpinner: React.FC<Props> = ({ text = null, style }) => {
```
Comment (larkox) [inline webapp/channels/src/components/widgets/loading/loading_spinner.tsx:15]: There is no need for the explicit typing. We prefer to have this as:
```
const ComponentName = ({
   prop1,
   ...
}: Props) => {
```

File: webapp/channels/src/components/widgets/loading/loading_spinner.tsx [hunk 4]
```diff
@@ -12,26 +12,22 @@ type Props = {
...
+    return (
+        <span
+            id='loadingSpinner'
+            className={'LoadingSpinner' + (text ? ' with-text' : '')}
```
Comment (larkox) [inline webapp/channels/src/components/widgets/loading/loading_spinner.tsx:19]: Out of the scope, but do you mind using `classNames` for this? You should be able to do something like this:
```suggestion
            className={classNames('LoadingSpinner', {'with-text': Boolean(text)})}
```

File: webapp/channels/src/components/widgets/loading/loading_spinner.tsx [hunk 5]
```diff
@@ -12,26 +12,22 @@ type Props = {
...
+            style={style}
+            data-testid='loadingSpinner'
+        >
+            <LocalizedIcon
+                className='fa fa-spinner fa-fw fa-pulse spinner'
+                component='span'
+                title={{ id: t('generic_icons.loading'), defaultMessage: 'Loading Icon' }}
```
Comment (larkox) [inline webapp/channels/src/components/widgets/loading/loading_spinner.tsx:26]: We want to get rid of the `t` function, so do you mind moving this outside of the component and define it with defineMessage?

Something like...

```
const IconTitle = defineMessage({id: 'generic_icons.loading', defaultMessage: 'Loading Icon'});
const LoadingSpinner: React.FC<Props> = ({ text = null, style }) => {
    ...
        title={IconTitle}
```
The other good thing about this is that you only define the string once, so you always pass the same object to the icon, avoiding potential re-renders down the line, and you put less strain in the GC, having to clean up the old message after every re-render.

**Review by larkox — CHANGES_REQUESTED** [review]
Thank you for the contribution! A few minor changes out of the scope, I hope you don't mind.

**Comment by larkox** [comment]
@Krishnakalani111 Also, see that there are a few lint and types errors in the checks. You should fix those too. Running `npm run fix` in the channels folder should fix most of the lint issues, and `npm run check-types` should let you know about the types errors.

**Comment by larkox** [comment]
@Krishnakalani111 Is this something you are still working on?

**Pull request closed without merging**